lib/Sema/ConstraintSystem.cpp

```diff
@@ -126,6 +126,33 @@ void ConstraintSystem::mergeEquivalenceClasses(TypeVariableType *typeVar1,
   // Merge nodes in the constraint graph.
   CG.mergeNodes(typeVar1, typeVar2);
 
+  // Let's try to de-duplicate any generic requirements
+  // associated with given type variable, if it represents
+  // a generic parameter.
+  if (typeVar1->getImpl().getGenericParameter()) {
+    auto requirements = CG.gatherConstraints(
+        typeVar1, ConstraintGraph::GatheringKind::EquivalenceClass,
+        [](Constraint *constraint) -> bool {
+          if (auto *locator = constraint->getLocator()) {
+            return locator->isLastElement(
+                       ConstraintLocator::TypeParameterRequirement) ||
+                   locator->isLastElement(
+                       ConstraintLocator::ConditionalRequirement);
+          }
+
+          return false;
+        });
+
+    llvm::SmallPtrSet<TypeBase *, 4> existingRequirements;
+    for (auto *req : requirements) {
+      auto constraintTy = req->getSecondType();
```

Comment:
Should we defensively check here whether the constraint itself has a second type? I feel like the locator checked above could be insufficient.

Reply:
We could `assert` but requirement is always a binary constraint from one type to another.

Reply:
Okay.